### PR TITLE
Issue #7242 Added Dockerfile for Docker container support

### PR DESCRIPTION
**Pull request by acwatkins**
I noticed a Vagrantfile. Docker is more lightweight/faster, especially for linux users. I have successfully built the firmware using this container.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM ubuntu:16.04
+ADD . /ardupilot
+WORKDIR /ardupilot
+
+RUN apt-get update && apt-get upgrade -y && apt-get dist-upgrade -y && apt-get install lsb-release sudo software-properties-common python-software-properties -y
+ENV USER=root
+RUN bash -c "Tools/scripts/install-prereqs-ubuntu.sh -y && apt-get install gcc-arm-none-eabi -y"
```
Comment (khancyr) [inline Dockerfile]: why use gcc-arm-none-eabi ? the install script should already have install the right toolchain !

**Comment by khancyr** [comment]
I am not at all a specialist on docker but here are my thoughts : 
will it not miss some config to load ardupilot Tools/ to the path ?
Why use Env ROOT ? we shouldn't work with root user
Shouldn't we expose default SILT port ?

**Comment by peterbarker** [comment]
On Fri, 17 Nov 2017, acwatkins wrote:

> I noticed a Vagrantfile. Docker is more lightweight/faster, especially for
> linux users. I have successfully built and flown the firmware using this
> container.

Awesome!

PR, please :-)

File: Dockerfile (hunk 1 above)
Comment (acwatkins) [reply Dockerfile]: I thought that as well, watching the script. Without this, gcc claimed it couldn't find it.

Comment (khancyr) [reply Dockerfile]: because you need to source the .profile file or at the toolchain location on the path !

**Comment by khancyr** [comment]
I think the idea is good but need a little more work to be as complet as Vagrant file !

**Comment by acwatkins** [comment]
@khancyr 

"Why use Env ROOT ? we shouldn't work with root user"
Docker images run as root by default, though you can set the user account. It would only be running for the exact command you are running and only be root in the container itself. Your script needs the USER environment variable set, which is the only reason that line is there. I originally made an ardupilot user and gave it full sudoer rights with no password, as there is no tty setup for password entry on a sudo command. At that point... I figure, what's the point.

"will it not miss some config to load ardupilot Tools/ to the path ? (with an entry point script ?)"
I thought your setup script setup the env with pathing.

"Shouldn't we expose default SILT port ?"
Could, you can also expose your entire network to the container at runtime, which is what I have been doing.

Example usage:

```
# Build the container
sudo docker build -t ardupilot .
# Run a command in the container
sudo docker run --name ardupilot --rm -it -v `pwd`:/ardupilot ardupilot bash -c "cd ArduCopter && make px4-v2"
```

The above builds the container, then runs a make command in the container, linking the local directory to a container directory and closing the container after it is all finished. If you want to pass your network to the container your would add --net=host to the run command.

**Comment by magicrub** [comment]
This was discussed in today's dev call. I'll test it with [Docker on Windows](https://www.docker.com/docker-windows). If there's no object, if it works I'll merge. We can discuss fixing/changing the user/root thing later if we want.

**Comment by magicrub** [comment]
here's another ArduPilot Docker resource we found..
https://hub.docker.com/r/gmyoungbloodparc/ardupilot-sitl/

**Comment by lvale** [comment]
@magicrub In fact there are quite a few Tom...

https://hub.docker.com/search/?isAutomated=0&isOfficial=0&page=1&pullCount=0&q=ardupilot&starCount=0

**Comment by gmorph** [comment]
Vagrant was used as it provides a host OS independent SITL environment.  Docker does not.  With docker its merely a container utilising whatever the host OS is.  This means it needs to be able to containerise a windows SITL environment - this is something we have struggled with in the past hence why Vagrant being able to virtual a linux one on any OS was the chosen solution.

**Comment by OXINARF** [comment]
@gmorph That's not exactly true. Docker can be used with a Linux base on Windows, but runs as a VM rather than a container (well, real containers are only available in Windows Server 2016, so everything on Windows are VMs).

File: Dockerfile (hunk 1 above)
Comment (EosBandi) [reply Dockerfile]: Quick note, AFAIK install-prereqs-ubuntu.sh tarball puts only a 32 bit version of gcc-arm-none-eabi to /opt not the 64bit version. Correct me if I'm wrong...

**Comment by peterbarker** [comment]
@asciich Can you test it works to satisfy your requirements?

**Comment by asciich** [comment]
@peterbarker Thx four your reply.

- First manual verification step would be: 
-- Check compiler version
-- Compare binary output with official nightly build. At least the size should more or less the same.
-- Run the compiled on a RC Quadrocopter should finally tell me if the compiler runs correclty.

Do you know if there are already some integration tests for the Firmware on real hardware?

**Comment by peterbarker** [comment]
On Wed, 28 Feb 2018, asciich wrote:

> Do you know if there are already some integration tests for the Firmware on real hardware?

We used to have a "Hardware in the Loop" mode, but I don't think we ever 
had tests based on it.

So... "no" :-)

**Comment by asciich** [comment]
@peterbarker  but I guess HITL would be the way to go.

I will try to set up SITL as a first step and then move to HITL (I see there some other benefits for me as well).

On the other hand this means it will take several days (too busy). I will write you again after I found a reasonable way to test it.

**Comment by davidbuzz** [comment]
LGTM.   I think it's fine to bring this in as-is, and any of the not-yet-implemented features can be added later as improvements in their own PR, if needed.

**Comment by jcs224** [comment]
Just thought I would revitalize this. There are lots of options on Docker Hub but I think it would by nice to have an officially-sanctioned container, even if it isn't perfect.

**Comment by peterbarker** [comment]
@jcs224 If we could get a small amount of documentation on how to run it (https://github.com/ArduPilot/ardupilot_wiki/edit/master/dev/source/docs/building-setup-linux.rst) we'd have more confidence merging it.

Just a paragraph or two explaining there's a docker file and a short command-line example of how to use it would be good.

**Comment by acwatkins** [comment]
I have updated based on comments and updated documentation.

**Comment by acwatkins** [comment]
https://github.com/acwatkins/ardupilot_wiki/pull/1

**Comment by peterbarker** [comment]
I've squished the commits together and fixed the commit message and pushed this into master - thanks!

**Pull request closed without merging**